### PR TITLE
#46565: Extend universal input/output support for ttnn::gather

### PR DESCRIPTION
**Pull request by mouliraj-mcw**
### Ticket
Link to Github Issue #46565

### Problem
`ttnn::gather` had no universal input/output story — any sharded output rejected with `TT_FATAL ("Output cannot be sharded")`, any `Layout::ROW_MAJOR` input rejected with `TT_FATAL ("input tensor must be in tiled format")`, and a sharded output requested without an explicit `shard_spec` would either fall through to a wrong layout or `TT_FATAL` downstream. Only TILE interleaved in / TILE interleaved out had a native path; every other (layout × sharding) combination was unsupported. The four existing kernels (`gather_{reader,writer}_single_row_{single,multi}_core.cpp`) were TILE-only by construction (face-tiled coordinate math), so there was no path forward for ROW_MAJOR without new kernels. Additionally, once sharded output-spec derivation was in place, a `COL_MAJOR` sharded input feeding a sharded-no-spec output silently flipped the output orientation to `ROW_MAJOR` — same family of bug as #48027 (transpose / permute / slice / repeat).

### What this PR does
- **Add composite-fallback predicate in `gather.cpp::detail`** — `needs_rm_irregular_composite` routes any RM B/W-sharded input *or* index whose last-dim is not `TILE_WIDTH`-aligned through an L1-interleaved TILE hop; every other (layout × sharding) combination flows natively. The composite arm round-trips **both** tensors through TILE to dodge the `noc_async_*_sharded` `pages_per_row` limitation on either side, and synthesizes a `shard_spec` via `generate_transpose_shard_spec` on the closing `to_memory_config` when the requested output is sharded-without-spec (the host op's spec derivation is bypassed on this path). Scoped to W-only since gather's output shape equals the index shape and never overshoots the buffer on H alone.
- **Add four new native RM kernels** — `gather_{reader,writer}_rm_single_row_{single,multi}_core.cpp` operate at stick (= row of W elements) granularity using `noc_async_*_sharded` so the same kernel handles interleaved, HEIGHT-sharded, and (regular) BLOCK/WIDTH-sharded buffers. The single-core variant is row-distributed (each core processes its assigned rows end-to-end), and the multi-core variant is column-distributed (each core owns a contiguous slice `[w_start, w_start + w_per_core)` of every H row). The reader performs the linear gather (`output[w] = input[index[w]]`) via new `get_value_from_stick` / `write_value_to_stick` helpers that reuse the existing datatype dispatch.
- **Pass per-shard `page_size_override` to the RM kernels** — `gather_program_factory.cpp::per_shard_page_size_bytes` computes `shard_W * elem_size` for BLOCK/WIDTH-sharded buffers (full row width otherwise) and feeds it as the `TensorAccessor` page-size override on the input, index, and output. Without this the kernel-side override would defeat the multi-shard split inside `noc_async_*_sharded` and corrupt regular-shape B/W I/O.
- **Synthesize output `shard_spec` in `compute_output_specs`** — when the user requests a sharded output without a spec, the device op derives one via `adjust_shard_spec_to_shape` (when the index tensor is sharded with a matching layout and rank) and falls back to `generate_transpose_shard_spec` otherwise. TILE rejects sub-tile adjusted shards and falls through to the fresh-spec branch.
- **Preserve `COL_MAJOR` shard orientation across synthesized output specs (#48025 pattern)** — both call sites of `generate_transpose_shard_spec` (native `compute_output_specs` fallback + composite arm's closing `to_memory_config`) now thread an explicit `orientation_hint` derived from the index tensor's `shard_spec` (falling back to the input tensor's on the composite path, since the staging hop discards `shard_spec` before the helper is called). Without this, a `COL_MAJOR` sharded input feeding a sharded-no-spec output silently flipped the output orientation to `ROW_MAJOR`. Same-layout sharded-to-sharded paths were unaffected because `adjust_shard_spec_to_shape` already preserves orientation.
- **Lift `TT_FATAL`s in `validate_on_program_cache_miss`** — drop the "Output cannot be sharded" rejection and the TILE-only layout rejection; add an input/index layout-consistency check (mixed TILE/RM unsupported because value/index/output streams are co-iterated stick-for-stick) and a tile-alignment check on user-supplied TILE sharded output `shard_spec`s.
- **Route `Layout::ROW_MAJOR` in `select_program_factory`** — TILE branches unchanged (`Wt > GATHER_WT_THRESHOLD` → `SingleRowMultiCore`, else `SingleRowSingleCore`); RM pivots on `W_index > GATHER_WT_THRESHOLD * TILE_WIDTH` (= 60 × 32 = 1920 elements) only, since the multi-core variant is column-distributed and `W_input` is mirrored in full to every core (no parallelism gain from `W_input`).
- **Skip `fill_implicit_tile_padding` on RM** — `pre_gather_transform_tensor` now checks `Layout::TILE` before calling `ttnn::fill_implicit_tile_padding`; RM has no tile-face padding so the helper would `TT_FATAL` ("FillPad should only be used for tile layout").
- **Cleanup in `post_gather_transform_tensor`** — dropped an unreachable ternary branch inside the `orig_rank > 4` arm (the inner `(orig_rank <= 4) ? ... : ...` was provably always taking the second branch).
- **Add comprehensive test suite (`test_gather_universal_io.py`, nightly)** — 81 cases organised into 13 lettered groups (A–N, slice/permute convention) covering: TILE interleaved + sharded I/O (bf16 + f32); TILE sharded-out without `shard_spec` (HEIGHT/WIDTH/BLOCK); TILE explicit shard configs (uneven splits and W-sharded inputs); RM interleaved single-core + multi-core + sharded I/O (bf16 + f32); RM sharded-out without spec (HEIGHT/BLOCK); RM sharded-in → sharded-no-spec-out (compute_output_specs derivation regression); irregular logical shapes (last-two not tile multiples) across {interleaved, HEIGHT, WIDTH, BLOCK} × {TILE, RM} (RM B/W routes composite); multi-batch / multi-channel sharded inputs; rank-5 sharded inputs; DRAM-sharded TILE inputs; default-mc regression guards (BLOCK_SHARDED inherits implicit out; RM W=49 fires composite); multi-core RM sharded variants (HEIGHT in/out, WIDTH in → interleaved, WIDTH in/out exercising `noc_async_write_sharded` with `w_per_core != shard_W`); pre-allocated `out=` on TILE and RM; `dim ∈ {0, 1, 2}` non-last-dim coverage (bf16 + f32); and `COL_MAJOR` shard-orientation preservation across native and composite synthesis paths. The main matrix runs **bf16** with **f32** mixed into Groups A, C, and L; bf8b is not covered natively because `fill_implicit_tile_padding` rejects BFP.

### Tests
All on Wormhole B0 (`tests/ttnn/nightly/unit_tests/operations/data_movement/test_gather_universal_io.py`):
- `test_gather_tile` — 18 cases (interleaved baselines, TILE HEIGHT/BLOCK sharded in/out, cross-layout, implicit inheritance × **bf16 + f32**) — **PASSED**
- `test_gather_tile_interleaved_to_sharded_no_spec` — 3 cases (HEIGHT/WIDTH/BLOCK × bf16) — **PASSED**
- `test_gather_rm` — 14 cases (RM interleaved single + multi-core, HEIGHT-sh, BLOCK-sh tile-aligned, cross-layout × **bf16 + f32**) — **PASSED**
- `test_gather_rm_interleaved_to_sharded_no_spec` — 2 cases (HEIGHT/BLOCK × bf16) — **PASSED**
- `test_gather_rm_sharded_in_sharded_no_spec_out` — 1 case (RM HEIGHT-sharded-in → HEIGHT no-spec-out, `compute_output_specs` derivation regression) — **PASSED**
- `test_gather_irregular_shapes_interleaved` — 4 cases (2 NoC-aligned irregular shapes × TILE/RM) — **PASSED**
- `test_gather_irregular_shapes_sharded` — 12 cases (2 shapes × HEIGHT/WIDTH/BLOCK × TILE/RM; RM B/W routes composite) — **PASSED**
- `test_gather_tile_explicit_sharded_inputs` — 2 cases (TILE HEIGHT-sharded 3-core uneven split, TILE WIDTH-sharded 4-core even split) — **PASSED**
- `test_gather_multi_batch_channel` — 2 cases (TILE N>1 height, TILE C>1 block) — **PASSED**
- `test_gather_higher_rank` — 2 cases (rank-5 TILE interleaved, rank-5 TILE HEIGHT-sharded) — **PASSED**
- `test_gather_higher_rank_to_sharded_no_spec` — 1 case (rank-5 TILE → HEIGHT-sharded out without spec) — **PASSED**
- `test_gather_dram_sharded_input` — 1 case (DRAM HEIGHT-sharded TILE input) — **PASSED**
- `test_gather_block_sharded_default_mc_no_longer_fatal` — 1 case (implicit-inheritance regression guard) — **PASSED**
- `test_gather_rm_irregular_width_composite_arm_fires` — 1 case (RM WIDTH-sharded W=49 → `needs_rm_irregular_composite`) — **PASSED**
- `test_gather_rm_multicore_sharded` — 3 cases (wide-W RM HEIGHT-sh in/out, wide-W RM WIDTH-sh → interleaved, wide-W RM WIDTH-sh in/out exercising `noc_async_write_sharded` with `w_per_core != shard_W`) — **PASSED**
- `test_gather_dim_variations` — 8 cases (`dim ∈ {0, 1, 2}` × bf16 + f32 on TILE/RM) — **PASSED**
- `test_gather_preallocated_output` — 2 cases (pre-allocated `out=` on TILE and RM, validates host-side optional-output pipeline) — **PASSED**
- `test_gather_col_major_orientation_preserved` — 4 cases (TILE same-layout fast path, TILE + RM cross-layout via hint, RM composite arm W=49 pre-hop snapshot; each asserts output `shard_spec.orientation == COL_MAJOR`) — **PASSED**

Total: **81 passed in ~19s**.

Pre-existing `tests/ttnn/unit_tests/operations/data_movement/test_gather.py` regression-clean (no changes to TILE-only paths).

One geometry deferred: `RM HEIGHT-sharded` with `W·E` not 16-byte aligned (e.g. `W = 97`, bf16). This is a `noc_async_*_sharded` helper-side limitation in `common.hpp` that affects `gather` / `slice` / `transpose` / `permute` identically — tracked under #47299. The parametrization for `(1, 1, 32, 97)` was removed from `_IRREGULAR_SHAPES` pending that upgrade and will be re-added when the helper fix lands.

### Notes for Reviewers
- The full gap analysis and routing tables (before/after) live on issue #46565 — recommended reading alongside the diff.
- Gather's pipeline adds **four new RM kernel files** because the pre-existing kernels were TILE-only with face-tiled coordinate math, and has **no `needs_sharded_output_reshard` equivalent** because the output shape always equals the **index** shape, so `compute_output_specs` can always seed a valid spec from the index tensor or via `generate_transpose_shard_spec`.
- Shared with the data-movement op family: `adjust_shard_spec_to_shape`, `generate_transpose_shard_spec` + its `orientation_hint` parameter (from `transpose/device/transpose_utils.hpp`), `tt::data_movement::common::noc_async_*_sharded` (from `common/kernels/common.hpp`), and the per-shard `page_size_override` pattern.
- The `needs_rm_irregular_composite` predicate retires when the cross-op helper fix in #47299 lands (same `pages_per_row` line; W-misalignment goes from composite to native).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mouliraj/gather_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mouliraj/gather_fix)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/gather/gather.cpp [hunk 1]
```diff
@@ -14,6 +14,33 @@
...
+
+namespace ttnn {
+namespace detail {
+
+// Fires if either input or index is RM B/W-sharded with non-tile-aligned W; the composite arm then
+// rounds-trips BOTH tensors through TILE to dodge the noc_async_*_sharded misread. Retires with #47299.
+inline bool needs_rm_irregular_composite(const ttnn::Tensor& input_tensor, const ttnn::Tensor& index_tensor) {
+    if (input_tensor.layout() != tt::tt_metal::Layout::ROW_MAJOR) {
+        return false;
+    }
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/gather/gather.cpp:27]: Typo: `rounds-trips` → `round-trips`

**Review by Aswinmcw — APPROVED** [review]

**Review by mgajewskiTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by VirdhatchaniKN — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_gather.py [hunk 2]
```diff
@@ -0,0 +1,976 @@
...
+import pytest
+import torch
+
+import ttnn
+
+from tests.ttnn.utils_for_testing import assert_with_pcc, assert_with_ulp
+
+_TTNN_TO_TORCH_DTYPE = {
+    ttnn.bfloat16: torch.bfloat16,
+    ttnn.float32: torch.float32,
```
Comment (KalaivaniMCW) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_gather.py:16]: do we not have full fp32 precision for gather ? is it an FPU operation ?

Comment (mouliraj-mcw) [reply tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_gather.py:16]: Gather is a pure data-movement op (dataflow kernels only), but when dim != -1 it calls ttnn.transpose in the host pipeline, and that transpose runs on the FPU.

File: ttnn/cpp/ttnn/operations/data_movement/gather/device/gather_device_operation.cpp [hunk 3]
```diff
@@ -15,9 +18,20 @@ constexpr uint32_t GATHER_WT_THRESHOLD = 60;
...
     const operation_attributes_t& /*operation_attributes*/, const tensor_args_t& tensor_args) {
-    // Calculate Wt to decide which program factory to use
     const auto input_tensor_shape = tensor_args.input_tensor.padded_shape();
     const auto input_index_tensor_shape = tensor_args.input_index_tensor.padded_shape();
+
+    if (tensor_args.input_tensor.layout() == Layout::ROW_MAJOR) {
+        // Multi-core splits over W_index only (each core owns a slice of every output row);
+        // W_input is mirrored in full to every core, so it gives no parallelism gain.
+        const uint32_t W_index = input_index_tensor_shape[-1];
+        const uint32_t rm_w_threshold = GATHER_WT_THRESHOLD * tt::constants::TILE_WIDTH;
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/data_movement/gather/device/gather_device_operation.cpp]: Nit: `constexpr`

**Review by amorrisonTT — APPROVED** [review]
